src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.PendingBranchesCollection.cs

```diff
@@ -0,0 +1,146 @@
+                    }
+                    _branchesToLabel.Free();
+                    _branchesToLabel = null;
+                }
+            }
+
+            internal ImmutableArray<PendingBranch> ToImmutable()
+            {
+                return _branchesToLabel is null ?
+                    _branches.ToImmutable() :
+                    ImmutableArray.CreateRange(this);
+            }
+
+            internal ArrayBuilder<PendingBranch> GetAndRemoveBranches(LabelSymbol? label)
+            {
+                ArrayBuilder<PendingBranch>? result;
+                if (label is null)
+                {
+                    result = _branches;
+                    _branches = ArrayBuilder<PendingBranch>.GetInstance();
+                }
+                else if (_branchesToLabel is { } && _branchesToLabel.TryGetValue(label, out result))
+                {
+                    _branchesToLabel.Remove(label);
+                }
+                else
+                {
+                    result = ArrayBuilder<PendingBranch>.GetInstance();
+                }
+                return result;
+            }
+
+            internal void Add(PendingBranch branch)
+            {
+                var label = branch.Label;
+                if (label is null)
+                {
+                    _branches.Add(branch);
+                }
+                else
+                {
+                    var branches = GetOrAddBranchesToLabel(label);
+                    branches.Add(branch);
+                }
+            }
+
+            internal void AddRange(PendingBranchesCollection collection)
+            {
+                _branches.AddRange(collection._branches);
+                if (collection._branchesToLabel is { })
+                {
+                    foreach (var pair in collection._branchesToLabel)
+                    {
+                        var branches = GetOrAddBranchesToLabel(pair.Key);
+                        branches.AddRange(pair.Value);
+                    }
+                }
+            }
+
+            private ArrayBuilder<PendingBranch> GetOrAddBranchesToLabel(LabelSymbol label)
+            {
+                if (_branchesToLabel is null)
+                {
+                    _branchesToLabel = PooledDictionary<LabelSymbol, ArrayBuilder<PendingBranch>>.GetInstance();
+                }
+                if (!_branchesToLabel.TryGetValue(label, out var branches))
+                {
+                    branches = ArrayBuilder<PendingBranch>.GetInstance();
+                    _branchesToLabel.Add(label, branches);
+                }
+                return branches;
+            }
+
+            public IEnumerator<PendingBranch> GetEnumerator()
+            {
+                return _branchesToLabel is null ?
+                    ((IEnumerable<PendingBranch>)_branches).GetEnumerator() :
+                    getEnumeratorCore();
+
+                IEnumerator<PendingBranch> getEnumeratorCore()
```

Comment:
Is it worthwhile to implement a struct enumerator here given that this change is being made for perf?

Reply:
If we did change `AsEnumerable()` to return a `struct`, we'd need to change `ToImmutable()` to iterate over the sorted sets of branches, so it would be somewhat involved. Perhaps we can revisit this if we see a perf issue.